docs/pages/advanced-topics/plugins-development.mdx

```diff
@@ -0,0 +1,310 @@
+	return "pod-hook"
+}
+
+func (p *podHook) Resource() client.Object {
+	return &corev1.Pod{}
+}
+```
+
+The `Name()` function defines the name of the hook which is used for logging purposes. The `Resource()` function returns the object you want to mutate. Besides those functions you can now define what actions you want to hook into inside vcluster's syncer:
+```
+type MutateCreateVirtual interface {
+	MutateCreateVirtual(ctx context.Context, obj client.Object) (client.Object, error)
+}
+
+type MutateUpdateVirtual interface {
+	MutateUpdateVirtual(ctx context.Context, obj client.Object) (client.Object, error)
+}
+
+type MutateDeleteVirtual interface {
+	MutateDeleteVirtual(ctx context.Context, obj client.Object) (client.Object, error)
+}
+
+type MutateGetVirtual interface {
+	MutateGetVirtual(ctx context.Context, obj client.Object) (client.Object, error)
+}
+
+type MutateCreatePhysical interface {
+	MutateCreatePhysical(ctx context.Context, obj client.Object) (client.Object, error)
+}
+
+type MutateUpdatePhysical interface {
+	MutateUpdatePhysical(ctx context.Context, obj client.Object) (client.Object, error)
+}
+
+type MutateDeletePhysical interface {
+	MutateDeletePhysical(ctx context.Context, obj client.Object) (client.Object, error)
+}
+
+type MutateGetPhysical interface {
+	MutateGetPhysical(ctx context.Context, obj client.Object) (client.Object, error)
+}
+```
+
+By implementing one or more of the above interfaces you will receive events from vcluster that allows you to mutate an outgoing or incoming object to vcluster.
+For example, to add an hook that adds a custom label to a pod, you can add the following code:
+```
+var _ hook.MutateCreatePhysical = &podHook{}
+
+func (p *podHook) MutateCreatePhysical(ctx context.Context, obj client.Object) (client.Object, error) {
+	pod, ok := obj.(*corev1.Pod)
+	if !ok {
+		return nil, fmt.Errorf("object %v is not a pod", obj)
+	}
+
+	if pod.Labels == nil {
+		pod.Labels = map[string]string{}
+	}
+	pod.Labels["created-by-plugin"] = "pod-hook"
+	return pod, nil
+}
+
+var _ hook.MutateUpdatePhysical = &podHook{}
+
+func (p *podHook) MutateUpdatePhysical(ctx context.Context, obj client.Object) (client.Object, error) {
+	pod, ok := obj.(*corev1.Pod)
+	if !ok {
+		return nil, fmt.Errorf("object %v is not a pod", obj)
+	}
+
+	if pod.Labels == nil {
+		pod.Labels = map[string]string{}
+	}
+	pod.Labels["created-by-plugin"] = "pod-hook"
+	return pod, nil
+}
+```
+
+Incoming objects into vcluster can be modified through the `MutateGetPhysical` or `MutateGetVirtual` which allows you to change how vcluster is retrieving objects from either the virtual or physical cluster.
+This can be useful if you don't want vcluster to change something you have mutated back for example.
+
+### Build and push your plugin
+
+Now you can run docker commands to build your container image and push it to the registry.
+`docker build -t your_org/vcluster-sync-all-configmaps . && docker push your_org/vcluster-sync-all-configmaps`
```

Comment:
perhaps make it a code block (```) so that is doesn't overlap with the right sidebar – ref. screenshot below

![image](https://github.com/loft-sh/vcluster/assets/7482025/79e941a5-4f56-4893-bfc5-694b75f08dd0)

Reply:
Done.